add Cryptid codecs

dhuseby opened the pull request:
This adds a number of multicodec values for various Cryptid projects ahead of the public release of the code. Most of the code is already available:
* [Multicid](https://github.com/cryptidtech/multicid)
  * Contains the VLAD implementation. [Spec](https://github.com/cryptidtech/provenance-specifications/blob/main/specifications/vlad.md)
* [Multisig](https://github.com/cryptidtech/multisig)
  * [Spec](https://github.com/cryptidtech/provenance-specifications/blob/main/specifications/multisig.md) 
* [Multikey](https://github.com/cryptidtech/multikey)
  * [Spec](https://github.com/cryptidtech/provenance-specifications/blob/main/specifications/multikey.md)
  *Contains the Nonce implementation as well. [Spec](https://github.com/cryptidtech/provenance-specifications/blob/main/specifications/nonce.md)

dhuseby commented:
The new Multisig and Multikey specs and implementations strictly follow:
* They MUST be in-band (with the value); not out-of-band (in context).
* They MUST avoid lock-in and promote extensibility.
* They MUST be compact and have a binary-packed representation.
* They MUST have a human-readable representation.

This cannot be said for the existing Varsig spec. It relies heavily on context to know how to decode the bytes of the digital signature. More importantly, if your implementation doesn't support a specific algorithm, there is no way to know how many bytes to skip over in the stream to skip over the Varsig. Multisig fixes this as an explicit goal.

rvagg commented:
Apologies for our slowness on this @dhuseby, both Volker and I are a bit stretched and you've given us a lot to review with your PRs and the cost of context switching to try and get our heads around all the new stuff in here isn't small.

Firstly, the one we tend to be most concerned about is anything that adds `ipld` - i.e. anything that might qualify for going into the "codec" portion of a CID.

1. If it's not intended to live in a CID then `ipld` is probably the wrong choice.
2. If it follows the same decoding and encoding process as an existing codec, then adding a new entry is probably not appropriate. It should be a new encoding format that's not already covered. We discourage using codec codes as a signalling mechanism for anything other than "how do I turn these bytes into the IPLD data model and back again".
3. If decoding bytes using this codec doesn't yield any "links" (concretely: CIDs), such that it can contribute to a larger graph, then `serialization` may be a better choice.

The new multi* entries will also require some additional thought and consideration. And due to the changes of the tags for existing entries we may need to pull in some folks who rely on those.

dhuseby commented:
@rvagg 

> If decoding bytes using this codec doesn't yield any "links" (concretely: CIDs), such that it can contribute to a larger graph, then serialization may be a better choice.

Ah...ok. These are a new IPLD data structure but I think you're right, `serialization` is probably correct here.

> The new multi* entries will also require some additional thought and consideration. And due to the changes of the tags for existing entries we may need to pull in some folks who rely on those.

This is a fair concern. I think to be safe, I'll just duplicate the overlap and assign new numbers.

Let me fix this.

dhuseby commented:
Thank you for taking time to review this. I changed the provenance-log and provenance-log-entry to be `serialization`. I changed a VLAD from `cid` to `vlad` since it is an identifier but not a CID. I restored the `varsig` entries to their original state and moved all of the `multisig` signature codecs to unique names and values to avoid collision. This should streamline the review and eliminate any push-back from repurposing.

dhuseby commented:
I also just corrected the `es384-msig` name error. I filed an issue for the same name error for the varsig: #348

dhuseby commented:
Also just cleaned up the `es521-msig` name error. I filed an issue for the same name error for varsig: #349

vmx commented:
I only had a quick look. We are generally pretty cautious adding anything into the number space which is encoded into a single byte as it is so limited (also some things that are already there shouldn't really, but that's hard to change). So if you would move your additions to a 2 bytes or higher section, that would be great.

dhuseby commented:
> So if you would move your additions to a 2 bytes or higher section, that would be great.

I can do that, but what about the `multikey`, `multisig`, `nonce`, and the `chacha20-poly1305`? I put all of these in the single byte space because they are fundamental types (or soon will be). I put `chacha20-poly1305` to be next to the existing `chacha` codecs.

I'm happy to move things elsewhere but I had reasons for putting these in the single byte space.

vmx commented:
> (or soon will be).

That's the point. At the moment we hardly add anything to the range of single bytes, the idea is to put things there that have wide adoption/are standardized. I know that it's annoying if it really gets used, that later on you don't really change to a different code, but that space is just so limited.

burdiyan commented:
FWIW, the word Multikey is currently being used in some of the upcoming W3C specs related to DID and Verifiable Credentials: https://www.w3.org/TR/vc-data-integrity/#multikey.

dhuseby commented:
My multikey predates the W3C use. There's no mention of multikey in this file yet. Plus the verifiable credentials spec has failed to gain adoption in key political and business areas.

dhuseby commented on table.csv:
```diff
@@ -517,9 +544,10 @@ json-jcs,                       ipld,           0xb601,         draft,      The
-bls-12381-g1-sig,               varsig,         0xd0ea,         draft,      G1 signature for BLS-12381-G2
-bls-12381-g2-sig,               varsig,         0xd0eb,         draft,      G2 signature for BLS-12381-G1
+bls12_381-g1-sig,               varsig,         0xd0ea,         draft,      G1 signature for BLS-12381-G2
+bls12_381-g2-sig,               varsig,         0xd0eb,         draft,      G2 signature for BLS-12381-G1
```
reverted

vmx reviewed (approved):
Those are lot of codecs, but as all of them are at least in the 2 byte range I don't see a reason not to accept them. We surely want people to use the multicodec table and experiment with things. Hence I approve it.

But before merging I'd like to see another approval from someone else, as it isn't just me making decisions.

rvagg commented on table.csv:
```diff
@@ -72,6 +72,7 @@ aes-192,                        key,            0xa1,           draft,      192-
 aes-256,                        key,            0xa2,           draft,      256-bit AES symmetric key
 chacha-128,                     key,            0xa3,           draft,      128-bit ChaCha symmetric key
 chacha-256,                     key,            0xa4,           draft,      256-bit ChaCha symmetric key
+chacha20-poly1305,              multikey,       0xa5,           draft,      ChaCha20_Poly1305 encryption scheme
```
One minor flag here - this is multi-byte so not a big deal, but adding this here doesn't leave any space for the plethora of authenticated encryption algorithms that it belongs with, I can easily imagine someone showing up and wanting to add a ton more but there being no space here. Maybe this belongs somewhere else where there's a bit more space to expand the AE list.

"multikey" seems a little limiting for this, but I don't have a better suggestion and I can't think of another way this might be consumed outside whatever you're doing.

dhuseby replied:
The multikey part is because Multikey objects can contain a chacha20-poly1305 symmetric key and tag. Multikeys can also contain an encrypted secret key that is encrypted using chacha20-poly1305.

I'm happy to move this to a 2-byte value.

rvagg replied:
It already encodes to a 2-byte varint so it's not a major blocker, it just doesn't allow room for additional AE schemes in this block, the only way it really fits is because it's next to chacha-* but it's actually quite different than plain chacha. Not a big deal, just trying to think ahead.

dhuseby replied:
Want me to move it? I'm happy to do that.

rvagg replied:
hey, how about we just bump it to `0xa00`, there's a huge block up there if we end up needing it, and 0xa00 encodes to the same number of varint bytes (2) as `0xa5`.

dhuseby replied:
ok, let me do that

rvagg commented:
I think I'm OK with this, it's just a lot of new tags being added, including single-entry ones: vlag, nonce, multikey. I'd be more comfortable if we didn't encourage more expansion of tags. But I don't have a hard reason to object.

dhuseby commented:
> including single-entry ones: vlag, nonce, multikey

At the risk of bikeshedding... vlad and cid tags should just be "identifier" or something generic like that. multikey and key could be combine into "key-like" or something generic like that. But that would require re-tagging a bunch of values which I'm sure is way worse of an outcome. The tags are all fundamental types that fit with the establish pattern IMO.

Thanks for the time you spent reviewing these.

dhuseby commented:
Thanks! Can we land this?

dhuseby commented:
I think all concerns have been addressed.

rvagg commented:
Sorry for the delays Dave, we're not well set up for dealing with complex requests like this.

vmx commented:
@dhuseby Thanks for your patience and all the changes compared to the original PR!

bumblefudge commented:
@rvagg @vmx Sorry for the late appearance, this crossed my inbox last week but I was at TPAC on a borrowed laptop and couldn't get here in time.

> My multikey predates the W3C use

This kinda feels like a "trust me bro" assertion.  The "W3C use" was based on early versions of juan benet's eternally-[unfinished/unpublished spec from 2016](https://github.com/ipfs/specs/issues/58) which seems consistent with references in the original commits in this org. In fact, I myself [removed `multikey` from the top-level readme in a 2023 PR](https://github.com/multiformats/multiformats/pull/65/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L20) exactly because I found discrepancies in various [proposed multikey variants](https://github.com/multiformats/multicodec/pull/125#discussion_r363243492) in the history of this org but no spec more canonical than JB's original issue (which matches _earlier versions_ of W3C usage, fwiw) to discriminate between the variants. The new spec seems perfectly defensible but I would consider giving it a suffix or renaming it to avoid confusion with the older `multikey` and its variants. If it doesn't get renamed here in this org, it might just cause headaches if this ever goes to IANA.

FWIW W3C DIDWG and VCWG do not use the term "multikey" anywhere, nor do they use by another name a similar CID-like multiformat; what they actually use (and specify internally) is a multibased raw-codec ULEB128 of public key bytes with the w3c-namespaced RDF/JSON-LD semantic [`publicKeyMultibase`](https://www.w3.org/TR/did-core/#dfn-publickeymultibase). Which is to say, the potential for confusion is, luckily, not as significant for the W3C term as for the multiformats' homegrown historical precedents. AFAIK none of the representatives from Maersk, GS1, or the Singaporean Trade Commission were lying about using `publicKeyMultibase` in prod last week at W3C TPAC?

Similarly, I believe there is some [multisig work ongoing elsewhere in the PL cinematic universe](https://github.com/ucan-wg/receipt?tab=readme-ov-file#depends-on) that I referred OP to when they asked about this a year ago, and never heard back about. That one goes by `varsig` rather than multisig and I'm not sure it's going to production or not, but could also be a source of avoidable confusion. Probably nothing but odd it wasn't mentioned in the PR.

dhuseby commented:
@bumblefudge I don't care what it is called. If "multikey" is a hotly contested name then let's go with "polykey" or "unikey" or "whateverkey" or "skeletonkey". I officially do not care. I just need these changes to get landed. So in my best effort to end the yak shaving and to celebrate Halloween/Diwali and our triumph of light over darkness/enlightenment over ignorance, I propose the use of "skeletonkey" to alleviate the concern.

dhuseby commented:
If Rust allowed for unicode type names, I'd even go so far as to call it "💀 key"

dhuseby commented:
> @dhuseby Thanks for your patience and all the changes compared to the original PR!

No worries! This is ~~Thunderdome~~ open source. 🪓

bumblefudge commented:
oh totally the sigil is yours! sorry it didn't land sooner.

dhuseby commented:
Thank you!
